homeassistant/components/nuki/__init__.py

```diff
@@ -1,3 +1,140 @@
+    DATA_BRIDGE,
+    DATA_COORDINATOR,
+    DATA_LOCKS,
+    DATA_OPENERS,
+    DEFAULT_PORT,
+    DEFAULT_TIMEOUT,
+    DOMAIN,
+    ERROR_STATES,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+NUKI_COMPONENTS = ["lock", "binary_sensor"]
+UPDATE_INTERVAL = timedelta(seconds=30)
+
+NUKI_SCHEMA = vol.Schema(
+    vol.All(
+        {
+            vol.Required(CONF_HOST): cv.string,
+            vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+            vol.Required(CONF_TOKEN): cv.string,
+        },
+    )
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {DOMAIN: vol.Schema(NUKI_SCHEMA)},
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+def _get_bridge_devices(bridge):
+    # pylint: disable=protected-access
+    locks = bridge._get_devices(0)
+    # pylint: disable=protected-access
+    openers = bridge._get_devices(2)
+    return locks, openers
+
+
+async def async_setup(hass, config):
+    """Set up the Nuki component."""
+
+    hass.data.setdefault(DOMAIN, {})
+
+    bridge = await hass.async_add_executor_job(
+        NukiBridge,
+        config[DOMAIN][CONF_HOST],
+        config[DOMAIN][CONF_TOKEN],
+        config[DOMAIN][CONF_PORT],
+        True,
+        DEFAULT_TIMEOUT,
+    )
+
+    try:
+        locks, openers = await hass.async_add_executor_job(_get_bridge_devices, bridge)
+    except RequestException as err:
+        raise exceptions.HomeAssistantError from err
+
+    async def async_update_data():
+        """Fetch data from Nuki bridge."""
+        try:
+            # Note: asyncio.TimeoutError and aiohttp.ClientError are already
+            # handled by the data update coordinator.
+            async with async_timeout.timeout(10):
+                for device in (
+                    hass.data[DOMAIN][DATA_LOCKS] + hass.data[DOMAIN][DATA_OPENERS]
+                ):
+                    for level in (False, True):
+                        try:
+                            await hass.async_add_executor_job(device.update, level)
+                        except RequestException:
+                            continue
+
+                        if device.state not in ERROR_STATES:
+                            break
+
+        except RequestException as err:
+            raise UpdateFailed(f"Error communicating with Bridge: {err}") from err
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        # Name of the data. For logging purposes.
+        name="nuki devices",
+        update_method=async_update_data,
+        # Polling interval. Will only be polled if there are subscribers.
+        update_interval=UPDATE_INTERVAL,
+    )
+
+    hass.data[DOMAIN] = {
+        DATA_COORDINATOR: coordinator,
+        DATA_BRIDGE: bridge,
+        DATA_LOCKS: locks,
+        DATA_OPENERS: openers,
+    }
+
+    # Fetch initial data so we have data when entities subscribe
+    await coordinator.async_refresh()
+
+    for component in NUKI_COMPONENTS:
+        hass.helpers.discovery.load_platform(component, DOMAIN, {}, config)
```

Comment:
This is deprecated, we need first update the component to config flow. Also, it mixes async and not sync code